OpenRA.Game/Primitives/FloatMatrix4x4.cs

```diff
@@ -0,0 +1,173 @@
+				Identity.M41, Identity.M42, Identity.M43, Identity.M44);
+		}
+
+		public static FloatMatrix4x4 CreateTranslation(float3 translation)
+		{
+			return new FloatMatrix4x4(
+				Identity.M11, Identity.M12, Identity.M13, Identity.M14,
+				Identity.M21, Identity.M22, Identity.M23, Identity.M24,
+				Identity.M31, Identity.M32, Identity.M33, Identity.M34,
+				translation.X, translation.Y, translation.Z, Identity.M44);
+		}
+
+		public static FloatMatrix4x4 operator *(FloatMatrix4x4 a, FloatMatrix4x4 b)
+		{
+			return new FloatMatrix4x4(
+				a.M11 * b.M11 + a.M21 * b.M12 + a.M31 * b.M13 + a.M41 * b.M14,
+				a.M12 * b.M11 + a.M22 * b.M12 + a.M32 * b.M13 + a.M42 * b.M14,
+				a.M13 * b.M11 + a.M23 * b.M12 + a.M33 * b.M13 + a.M43 * b.M14,
+				a.M14 * b.M11 + a.M24 * b.M12 + a.M34 * b.M13 + a.M44 * b.M14,
+				a.M11 * b.M21 + a.M21 * b.M22 + a.M31 * b.M23 + a.M41 * b.M24,
+				a.M12 * b.M21 + a.M22 * b.M22 + a.M32 * b.M23 + a.M42 * b.M24,
+				a.M13 * b.M21 + a.M23 * b.M22 + a.M33 * b.M23 + a.M43 * b.M24,
+				a.M14 * b.M21 + a.M24 * b.M22 + a.M34 * b.M23 + a.M44 * b.M24,
+				a.M11 * b.M31 + a.M21 * b.M32 + a.M31 * b.M33 + a.M41 * b.M34,
+				a.M12 * b.M31 + a.M22 * b.M32 + a.M32 * b.M33 + a.M42 * b.M34,
+				a.M13 * b.M31 + a.M23 * b.M32 + a.M33 * b.M33 + a.M43 * b.M34,
+				a.M14 * b.M31 + a.M24 * b.M32 + a.M34 * b.M33 + a.M44 * b.M34,
+				a.M11 * b.M41 + a.M21 * b.M42 + a.M31 * b.M43 + a.M41 * b.M44,
+				a.M12 * b.M41 + a.M22 * b.M42 + a.M32 * b.M43 + a.M42 * b.M44,
+				a.M13 * b.M41 + a.M23 * b.M42 + a.M33 * b.M43 + a.M43 * b.M44,
+				a.M14 * b.M41 + a.M24 * b.M42 + a.M34 * b.M43 + a.M44 * b.M44);
+		}
+
+		public static float4 operator *(FloatMatrix4x4 a, float4 b)
+		{
+			return new float4(
+				a.M11 * b.X + a.M21 * b.Y + a.M31 * b.Z + a.M41 * b.W,
+				a.M12 * b.X + a.M22 * b.Y + a.M32 * b.Z + a.M42 * b.W,
+				a.M13 * b.X + a.M23 * b.Y + a.M33 * b.Z + a.M43 * b.W,
+				a.M14 * b.X + a.M24 * b.Y + a.M34 * b.Z + a.M44 * b.W);
+		}
+
+		public static bool operator ==(FloatMatrix4x4 me, FloatMatrix4x4 other)
+		{
+			return
+				me.M11 == other.M11 && me.M12 == other.M12 && me.M13 == other.M13 && me.M14 == other.M14 &&
+				me.M21 == other.M21 && me.M22 == other.M22 && me.M23 == other.M23 && me.M24 == other.M24 &&
+				me.M31 == other.M31 && me.M32 == other.M32 && me.M33 == other.M33 && me.M34 == other.M34 &&
+				me.M41 == other.M41 && me.M42 == other.M42 && me.M43 == other.M43 && me.M44 == other.M44;
+		}
+
+		public static bool operator !=(FloatMatrix4x4 me, FloatMatrix4x4 other) { return !(me == other); }
+
+		public override int GetHashCode() { return M11.GetHashCode() ^ M22.GetHashCode() ^ M33.GetHashCode() ^ M44.GetHashCode(); }
+
+		public bool Equals(FloatMatrix4x4 other) { return other == this; }
+		public override bool Equals(object obj) { return obj is FloatMatrix4x4 && Equals((FloatMatrix4x4)obj); }
+
+		public override string ToString()
+		{
+			return
+				"[" + M11 + " " + M12 + " " + M13 + " " + M14 + "],[" +
```

Comment:
This seems to open more `[` than it closes.

Reply:
Oh! thanks, copied that from Int32Matrix4x4. So both need a fix then :D